astropy/io/fits/fitstime.py

```diff
@@ -0,0 +1,289 @@
+                        'Table column {} has a FITS recognized time scale value {}. '
+                        'However, since it is not a valid astropy time scale, '
+                        'it will not be converted to astropy Time.'.format(idx, scale),
+                        AstropyUserWarning)
+                # Non-linear coordinate types have "4-3" form and are not time coordinates
+                elif TCTYP_RE_TYPE.match(scale[:5]) and TCTYP_RE_ALGO.match(scale[5:]):
+                    pass
+                else:
+                    warnings.warn(
+                        'Table column {} has a coordinate type {} which is either '
+                        'an unrecognized local time scale or a linear coordinate-'
+                        'type'.format(idx, scale),
+                        AstropyUserWarning)
+                time_columns.pop(idx)
+            continue
+
+        if (unit is not None and unit in FITS_TIME_UNIT) or 'pos' in time_col:
+            continue
+
+        # Other cases are yet to be checked and will be done when reading starts
+        # This includes checking TUNITn
+        time_columns.pop(idx)
+
+
+def _convert_time_columns(table, global_info, time_columns):
+    """
+    Convert time columns to astropy Time columns.
+
+    Parameters
+    ----------
+    table : `~astropy.table.Table`
+        The table whose time columns are to be converted.
+    global_info : dict
+        Global time reference coordinate information
+    time_columns : dict
+        Column-specific time information
+    """
+
+    # The code might fail while attempting to read FITS files not written by astropy.
+
+    # Read in Global Informational keywords as Time
+    for key in global_info:
+        # FITS uses a subset of ISO-8601 for several time-related keywords,
+        # such as DATE-xxx
+        if key.startswith('date'):
+            if key not in table.meta:
+                try:
+                    table.meta[key.upper()] = Time(global_info[key], scale=global_info['scale'],
+                                                   precision=(lambda x: len(x.split('.')[1])
+                                                   if '.' in x else 0)(global_info[key]))
+                except ValueError:
+                    table.meta[key.upper()] = global_info[key]
+
+        # MJD-xxx
+        elif key.startswith('mjd-'):
+            if key not in table.meta:
+                try:
+                    table.meta[key.upper()] = Time(global_info[key],
+                                                   scale=global_info['scale'],
+                                                   format='mjd')
+                except ValueError:
+                    table.meta[key.upper()] = global_info[key]
+
+    # Read in time coordinate columns as Time
+    for idx, time_col in time_columns.items():
+        time_colname = table.colnames[idx - 1]
+        table[time_colname] = Time(table[time_colname][...,0], table[time_colname][...,1],
```

Comment:
Check shape of `table[time_colname]`.  Must have `.shape[-1] == 2` and `.ndim > 1`.